### PR TITLE
Support run mr_test on Public Cloud

### PR DESCRIPTION
**Pull request by lilyeyes**
Support run mr_test on Public Cloud: Azure

NOTE:
1. about the PC image:  https://dist.suse.de/ibs/SUSE:/${product}:/GA:/TEST/images/${product}-SAP-BYOS.x86_64.*-Azure-.*
    (I did not include public image this time as `qesapdeployment` still failed on create instances)
2. about SLE 15: we don't test this product, see TEAM-6798 for more details
3. about cleanup: test module "ssh_interactive_end" does the cleanup
4. about the failure: all issues are known and tracked by https://progress.opensuse.org/issues/119806

- Related ticket: https://jira.suse.com/browse/TEAM-6798
- Needles: NA
- Related MR:
  https://gitlab.suse.de/qa-css/openqa_ha_sap/-/merge_requests/386
  https://gitlab.suse.de/qa-maintenance/metadata/-/merge_requests/727
- Needles: NA
- Verification run: Flavor: Azure-SAP-BYOS-Incidents-saptune

    Maintenance: SLE 15 SP4 SAP Incidents (all passed): https://openqaworker15.qa.suse.cz/tests/overview?distri=sle&version=15-SP4&build=%3A26445%3Arelease-notes-sles-for-sap&groupid=21
    Maintenance: SLE 15 SP3 SAP Incidents (all failure can be tracked by poo#119806): https://openqaworker15.qa.suse.cz/tests/overview?distri=sle&version=15-SP3&build=%3A26446%3Arelease-notes-sles-for-sap&groupid=22
    Maintenance: SLE 15 SP1 SAP Incidents (all failure can be tracked by poo#119806): https://openqaworker15.qa.suse.cz/tests/overview?distri=sle&version=15-SP1&build=%3A26742%3Asupportutils&groupid=24
    Maintenance: SLE 12 SP4 SAP Incidents (all failure can be tracked by poo#119806):       https://openqaworker15.qa.suse.cz/tests/overview?distri=sle&version=12-SP4&build=%3A26650%3ASUSEConnect&groupid=29

    Those tests are successfully triggered and test results are as expected before but I did re-trigger and they are failed as incident repos were removed:
  Maintenance: SLE 15 SP2 SAP Incidents: https://openqaworker15.qa.suse.cz/tests/overview?distri=sle&version=15-SP2&build=%3A26242%3Asupportutils-plugin-ha-sap&groupid=23
  Maintenance: SLE 12 SP5 SAP Incidents: https://openqaworker15.qa.suse.cz/tests/overview?distri=sle&version=12-SP5&build=%3A26243%3Asupportutils-plugin-ha-sap&groupid=28

    This test case is the verification run for LTP as I revised other team's code (passed, did not introduce test code regression): https://openqa.suse.de/tests/9916049

### CODE REVIEW

File: tests/sles4sap/saptune/mr_test.pm
Comment (mpagot) [inline tests/sles4sap/saptune/mr_test.pm:27]: running

File: tests/publiccloud/mr_test_setup_env.pm
Comment (alvarocarvajald) [inline tests/publiccloud/mr_test_setup_env.pm:6]: Should this module be located in `tests/publiccloud/` or in `tests/sles4sap/publiccloud`?

File: tests/publiccloud/mr_test_setup_env.pm
Comment (alvarocarvajald) [inline tests/publiccloud/mr_test_setup_env.pm:17]: Any reason to leave these two lines commented?

File: tests/publiccloud/mr_test_setup_env.pm
Comment (alvarocarvajald) [inline tests/publiccloud/mr_test_setup_env.pm:52]: s/ombit/ommit/

File: tests/sles4sap/saptune/mr_test.pm
Comment (alvarocarvajald) [reply tests/sles4sap/saptune/mr_test.pm:27]: More than they typo in `running`, I'd say one of the 2 instances of *ing needs to go. It could be either:
1. Do not reboot if testing in public cloud instance,
2. Do not reboot if test is running in public cloud instance.

File: tests/sles4sap/saptune/mr_test.pm
Comment (alvarocarvajald) [inline tests/sles4sap/saptune/mr_test.pm:31]: Suggestion:
```
$self->reboot unless get_var('PUBLIC_CLOUD_SLES4SAP');
```

File: tests/sles4sap/saptune/mr_test.pm
Comment (alvarocarvajald) [inline tests/sles4sap/saptune/mr_test.pm:61]: Suggestion:
```
assert_script_run "curl -sk $tarball | tar zxf - --strip-components 1" unless get_var('PUBLIC_CLOUD_SLES4SAP');
```

File: tests/publiccloud/mr_test_setup_env.pm
Comment (lilyeyes) [reply tests/publiccloud/mr_test_setup_env.pm:6]: Good question. I had considered putting this file in "tests/publiccloud/" but it it is `mr_test` specific environment setup so I put it in "tests/sles4sap/publiccloud".

File: tests/publiccloud/mr_test_setup_env.pm
Comment (lilyeyes) [reply tests/publiccloud/mr_test_setup_env.pm:17]: Thanks, deleted.

File: tests/sles4sap/saptune/mr_test.pm
Comment (lilyeyes) [reply tests/sles4sap/saptune/mr_test.pm]: Done.

File: tests/sles4sap/saptune/mr_test.pm
Comment (lilyeyes) [reply tests/sles4sap/saptune/mr_test.pm]: Done.

File: tests/publiccloud/mr_test_setup_env.pm
Comment (lilyeyes) [reply tests/publiccloud/mr_test_setup_env.pm]: Done.

File: lib/publiccloud/ssh_interactive.pm
Comment (lilyeyes) [reply lib/publiccloud/ssh_interactive.pm]: Done. Thanks. Those code is moved from "tests/publiccloud/prepare_instance.pm" so I did not check them.

File: lib/mr_test_run.pm
Comment (lilyeyes) [reply lib/mr_test_run.pm]: Done.

File: lib/mr_test_run.pm
Comment (lilyeyes) [reply lib/mr_test_run.pm]: Done.

File: tests/sles4sap/saptune/mr_test.pm
Comment (lilyeyes) [reply tests/sles4sap/saptune/mr_test.pm]: Done. Thank you all :)

**Comment by dzedro** [comment]
I don't think placing mr_test_run into lib/ is good idea.

**Comment by lilyeyes** [comment]
> I don't think placing mr_test_run into lib/ is good idea.

I put it here as it is a package and is used in `./lib/mr_test_lib.pm:use mr_test_run qw(get_notes get_solutions);`
Should I put it in `tests/sles4sap/saptune/` ?

**Comment by dzedro** [comment]
> > I don't think placing mr_test_run into lib/ is good idea.
> 
> I put it here as it is a package and is used in `./lib/mr_test_lib.pm:use mr_test_run qw(get_notes get_solutions);` Should I put it in `tests/sles4sap/saptune/` ?

Why should it be in lib/,  when it's not library ?

**Comment by lilyeyes** [comment]
> 
I was not sure if putting it in "tests/sles4sap/saptune/" can be `use mr_test_run` so I put it in `lib`.
Let me try put it in  "tests/sles4sap/saptune/"

**Comment by lilyeyes** [comment]
> > > I don't think placing mr_test_run into lib/ is good idea.
> > 
> > 
> > I put it here as it is a package and is used in `./lib/mr_test_lib.pm:use mr_test_run qw(get_notes get_solutions);` Should I put it in `tests/sles4sap/saptune/` ?
> 
> Why should it be in lib/, when it's not library ?

I was not sure if putting it in "tests/sles4sap/saptune/" can be use mr_test_run so I put it in lib.
Let me try put it in "tests/sles4sap/saptune/"

**Comment by lilyeyes** [comment]
@dzedro  Done, moved it to `tests/sles4sap/saptune/` ! VR: https://openqaworker15.qa.suse.cz/tests/67256

**Review by alvarocarvajald — APPROVED** [review]
LGTM.

**Comment by dzedro** [comment]
> @dzedro Done, moved it to `tests/sles4sap/saptune/` ! VR: https://openqaworker15.qa.suse.cz/tests/67256

Same applies also for `lib/publiccloud/ssh_interactive.pm`, as it's test not library.

**Comment by lilyeyes** [comment]
> > @dzedro Done, moved it to `tests/sles4sap/saptune/` ! VR: https://openqaworker15.qa.suse.cz/tests/67256
> 
> Same applies also for `lib/publiccloud/ssh_interactive.pm`, as it's test not library.

This file' `Maintainer: qa-c@suse.de` I would like to not touch it.

**Comment by dzedro** [comment]
> > > @dzedro Done, moved it to `tests/sles4sap/saptune/` ! VR: https://openqaworker15.qa.suse.cz/tests/67256
> > 
> > 
> > Same applies also for `lib/publiccloud/ssh_interactive.pm`, as it's test not library.
> 
> This file' `Maintainer: qa-c@suse.de` I would like to not touch it.

Sorry, I didn't look good enough, it's library.

File: lib/mr_test_lib.pm
Comment (asmorodskyi) [inline lib/mr_test_lib.pm:68]: be aware that `ssh_interactive_end` has test flags `fatal` and `publiccloud_multi_module` ( see  https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/publiccloud/ssh_interactive_end.pm#L23 )  . which means that cleanup will happen only in case of module failure  see https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/publiccloud/basetest.pm#L120 . I suspect that this is not what you expecting

Comment (lilyeyes) [reply lib/mr_test_lib.pm:68]: > be aware that `ssh_interactive_end` has test flags `fatal` and `publiccloud_multi_module` ( see https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/tests/publiccloud/ssh_interactive_end.pm#L23 ) . which means that cleanup will happen only in case of module failure see https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/publiccloud/basetest.pm#L120 . I suspect that this is not what you expecting

When all test module passed, cleanup is done: https://openqaworker15.qa.suse.cz/tests/65672#step/ssh_interactive_end/28
When one of the test module failed, cleanup is done also: https://openqaworker15.qa.suse.cz/tests/67256#step/ssh_interactive_end/28
https://openqaworker15.qa.suse.cz/tests/55628#step/ssh_interactive_end/4

File: lib/publiccloud/ssh_interactive.pm
Comment (asmorodskyi) [inline lib/publiccloud/ssh_interactive.pm:128]: I think it make sense . @grisu48 , @pdostal  WDYT ?

File: lib/publiccloud/ssh_interactive.pm
Comment (asmorodskyi) [inline lib/publiccloud/ssh_interactive.pm:153]: I assume it was not intentional , you simply haven't re-based for quite some time so you dropping here https://github.com/os-autoinst/os-autoinst-distri-opensuse/commit/c5b5e1c227ac9908b08439d73345fb057dc62ca2 which was merged recently

File: tests/sles4sap/publiccloud/mr_test_setup_env.pm
Comment (asmorodskyi) [inline tests/sles4sap/publiccloud/mr_test_setup_env.pm:35]: you don't have to do this , @mimi1vx  found a better way , see https://github.com/os-autoinst/os-autoinst-distri-opensuse/commit/02f132014f2d87725d5bf2f29bb40b56764da3a8

File: tests/sles4sap/publiccloud/mr_test_setup_env.pm
Comment (asmorodskyi) [inline tests/sles4sap/publiccloud/mr_test_setup_env.pm:41]: why you not using git to checkout git project ?

File: tests/sles4sap/publiccloud/mr_test_setup_env.pm
Comment (asmorodskyi) [inline tests/sles4sap/publiccloud/mr_test_setup_env.pm:55]: in cloud we have plenty of cases where username is provider specific . For Azure it is `azure` . https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/publiccloud/azure_client.pm#L18 So unless you really know what you are doing I would not touch username . Every provider perfectly know what needs to be used here

File: lib/mr_test_lib.pm
Comment (asmorodskyi) [reply lib/mr_test_lib.pm:68]: ok I didn't realize that `ssh_interactive_end` is not ancestor of basetest. You are correct , ignore my message

File: tests/sles4sap/publiccloud/mr_test_setup_env.pm
Comment (asmorodskyi) [reply tests/sles4sap/publiccloud/mr_test_setup_env.pm:41]: because you will copy it into the cloud as tarball. make sense

File: tests/sles4sap/saptune/mr_test.pm
Comment (asmorodskyi) [inline tests/sles4sap/saptune/mr_test.pm:62]: can you elaborate why you need to add root to PATH ?

File: tests/sles4sap/saptune/mr_test.pm
Comment (asmorodskyi) [inline tests/sles4sap/saptune/mr_test.pm:101]: if you need to cleanup just call it directly no need to load whole module just to achieve this

File: tests/sles4sap/saptune/mr_test_run.pm
Comment (asmorodskyi) [inline tests/sles4sap/saptune/mr_test_run.pm:40]: I would keep `my ($self) = @_;` out nevertheless

File: tests/sles4sap/saptune/mr_test_run.pm
Comment (asmorodskyi) [inline tests/sles4sap/saptune/mr_test_run.pm:37]: rebooting in clouds is tough and complicated you right , but if you need this reboot and if you doing it in other tests does it really make sense simply ignore this ? Is your test still valid if you not doing reboot ? In some cases (e.g. kernel update)  reboot is vital and without it further testing does not make any sense . While reboot is complex and error prone for public cloud tests it is not impossible

File: lib/publiccloud/ssh_interactive.pm
Comment (grisu48) [reply lib/publiccloud/ssh_interactive.pm:128]: A prima vista, yes. I would like to see some verification runs on our maintenance test runs though.

File: tests/sles4sap/saptune/mr_test.pm
Comment (lilyeyes) [reply tests/sles4sap/saptune/mr_test.pm:62]: Under non public cloud env the $HOME is root; So in order to keep the same testing environment with non public cloud env, I added "/root" to $PATH.

File: tests/sles4sap/publiccloud/mr_test_setup_env.pm
Comment (lilyeyes) [reply tests/sles4sap/publiccloud/mr_test_setup_env.pm]: I chose "bernhard" as it is a very common/popular user used in openQA;
I tried "cloudadmin" but it reports invalid user: https://openqaworker15.qa.suse.cz/tests/67494#step/mr_test_setup_env/37
So I prefer to keep using this "bernhard".

File: tests/sles4sap/saptune/mr_test.pm
Comment (lilyeyes) [reply tests/sles4sap/saptune/mr_test.pm]: Done.

File: tests/sles4sap/saptune/mr_test_run.pm
Comment (lilyeyes) [reply tests/sles4sap/saptune/mr_test_run.pm]: Done.

File: tests/sles4sap/saptune/mr_test_run.pm
Comment (lilyeyes) [reply tests/sles4sap/saptune/mr_test_run.pm:39]: I have asked in team no concrete answer yet, if it is essential I will open another JIRA ticket to track this comment.
Thank you so much for raising it up.

**Comment by lilyeyes** [comment]
> It seems the move of `mr_test_run` from `./lib/` to `./tests/` broke the CI tests. I guess there are still modules doing `use mr_test_run;`

It seems a common or network issue, a number of PRs are failed on similar failure.

File: schedule/sles4sap/sles4sap_gnome_saptune.yaml
Comment (mpagot) [inline schedule/sles4sap/sles4sap_gnome_saptune.yaml:9]: How this one is used? How it is injected in the deployment process?  It is not part of sles4sap_azure_generic.yaml?

File: lib/publiccloud/ssh_interactive.pm
Comment (lilyeyes) [reply lib/publiccloud/ssh_interactive.pm:128]: Hi @grisu48 & @asmorodskyi,
I have revised code according to you comments and did VR, please help to check again, thanks :)
VR-1 for `mr_test`:
   passed as expected: https://openqaworker15.qa.suse.cz/tests/67499#
   failed as expected: http://openqaworker15.qa.suse.cz/tests/67504#
  
VR-2 for `qem` test cases
   publiccloud_nfs passed: http://openqa.suse.de/tests/9961054
   publiccloud_smoketests_gen1 passed: http://openqa.suse.de/tests/9961043
   
VR-3 for `post fail cleanup` on purpose
    sles4sap_gnome_saptune_delete_rename: http://openqaworker15.qa.suse.cz/t67504

File: tests/sles4sap/publiccloud/mr_test_setup_env.pm
Comment (lilyeyes) [reply tests/sles4sap/publiccloud/mr_test_setup_env.pm]: Deleted. Thanks for the info :+1:

File: lib/publiccloud/ssh_interactive.pm
Comment (lilyeyes) [reply lib/publiccloud/ssh_interactive.pm]: > I assume it was not intentional , you simply haven't re-based for quite some time so you dropping here [c5b5e1c](https://github.com/os-autoinst/os-autoinst-distri-opensuse/commit/c5b5e1c227ac9908b08439d73345fb057dc62ca2) which was merged recently

Thank you so much for this. Done.

**Review by mpagot — APPROVED** [review]
LGTM

File: schedule/sles4sap/sles4sap_gnome_saptune.yaml
Comment (lilyeyes) [reply schedule/sles4sap/sles4sap_gnome_saptune.yaml:9]: > How this one is used? How it is injected in the deployment process? It is not part of sles4sap_azure_generic.yaml?

Good question, I remember if not set "PUBLIC_CLOUD_RESOURCE_GROUP" test case will be failed. But I can not find the failed openQA run currently.

Comment (lilyeyes) [reply schedule/sles4sap/sles4sap_gnome_saptune.yaml:9]: Maybe the new split qesapdeployment code don't need this variable.
I used 'tests/publiccloud/prepare_instance.pm" to create instance before. Let me try to delete it.

**Review by asmorodskyi — APPROVED** [review]
LGTM if cloning of this jobs will pass : 
https://openqa.suse.de/tests/9957061
https://openqa.suse.de/tests/9957062
https://openqa.suse.de/tests/9957059

File: schedule/sles4sap/sles4sap_gnome_saptune.yaml
Comment (lilyeyes) [reply schedule/sles4sap/sles4sap_gnome_saptune.yaml]: Deleted. It is useless now: https://openqaworker15.qa.suse.cz/tests/67509#step/ssh_interactive_end/28

**Comment by lilyeyes** [comment]
> LGTM if cloning of this jobs will pass : https://openqa.suse.de/tests/9957061 https://openqa.suse.de/tests/9957062 https://openqa.suse.de/tests/9957059

All passed:
http://openqa.nue.suse.com/tests/9961632
http://openqa.nue.suse.com/tests/9961644
http://openqa.nue.suse.com/tests/9961655

File: tests/sles4sap/publiccloud/mr_test_setup_env.pm
Comment (alvarocarvajald) [reply tests/sles4sap/publiccloud/mr_test_setup_env.pm]: Yes, those commands are running in the jumphost, not in the cloud infra. The command right before that one gives it away: https://openqaworker15.qa.suse.cz/tests/67494#step/mr_test_setup_env/33

**Review by alvarocarvajald — APPROVED** [review]
LGTM

**Comment by lilyeyes** [comment]
> > It seems the move of `mr_test_run` from `./lib/` to `./tests/` broke the CI tests. I guess there are still modules doing `use mr_test_run;`
> 
> It seems a common or network issue, a number of PRs are failed on similar failure.

You are right the move also introduced CI errors, so I revised "Makefile" and all checks are passed (finally) now.
Please help to check and merge.

**Comment by lilyeyes** [comment]
> > It seems the move of `mr_test_run` from `./lib/` to `./tests/` broke the CI tests. I guess there are still modules doing `use mr_test_run;`
> 
> It seems a common or network issue, a number of PRs are failed on similar failure.

You are right the move also introduced CI errors, so I revised "Makefile" and all checks are passed (finally) now.
Please help to check and merge.

**Comment by alvarocarvajald** [comment]
@asmorodskyi are you guys OK with us merging this?

**Comment by asmorodskyi** [comment]
> @asmorodskyi are you guys OK with us merging this?

yes Looks Good To ME

**Pull request merged**
